### ISSUE TITLE
Issue with package including a py.typed file

### ISSUE THREAD

**Issue by tjsmart** [microsoft/pylance-release#2965]
tldr: I am experiencing an unexpected tab completion result when I install a package using `pip install -e` which contains a `py.typed` file. Minimal working example shown below.

## Environment data

-   Language Server version: pylance v2022.6.30
-   OS and version: Ubuntu 18.04.6 LTS
-   Python version (& distribution if applicable, e.g. Anaconda): Python 3.8.13

## Code Snippet

Here is a minimal working example:
[example.zip](https://github.com/microsoft/pylance-release/files/8976845/example.zip)

## Repro Steps

1. Create a folder with the contents of example.zip. Should look like this:

```
example
├── baz
│   └── demo.py
└── foo
    ├── foo
    │   ├── _bar.py
    │   ├── __init__.py
    │   └── py.typed
    └── setup.py
```

2. Open the folder `baz`, using vscode. (I do think exclusively opening the folder baz is important, but I'm not certain.)

```bash
code baz
```

3. Create a virtual environment and install an editable version of the package foo.

```bash
python3.8 -m virtualenv venv
. venv/bin/activate
pip install -e ../foo
```

The contents of `pip list` should show something like this:

```            
Package    Version Editable project location
---------- ------- -------------------------
foo        0.0.0   /home/tsmart/example/foo
pip        22.1.2
setuptools 62.3.4
wheel      0.37.1
```

4. Select `./venv/bin/python` as the python interpreter and reload the window.

5. Open the file `demo.py` and try typing `from foo import `, tab completion will show:

![image](https://user-images.githubusercontent.com/40041862/175553499-62152a60-58ad-4a8b-97ea-2c1443652e45.png)

***But I expected to see the variable `a` in the output window for tab completion.***

6. Interesting thing, if I remove the file `example/foo/foo/py.typed` I get the result I expected to also see in step 5 (don't forget to reload the window):

![image](https://user-images.githubusercontent.com/40041862/175553923-177bb9c4-20ab-4ae5-8676-24ed3634d043.png)

## Logs

Here is the log from step 5 above: 
[bad_step5_log.txt](https://github.com/microsoft/pylance-release/files/8976891/bad_step5_log.txt)

Here is the log from step 6 above:
[good_step6_log.txt](https://github.com/microsoft/pylance-release/files/8976894/good_step6_log.txt)

**Comment by tjsmart** [comment]
Update: Should of tried this before submitting but I do not believe the editable install is relevant to the issue. I can reproduce the issue using a normal install. Although, for convenience I would still suggest doing an editable install to reproduce the issue.

Note: The `setup.py` I included in my example won't include any data files, so `py.typed` will be missing it if you simply `pip install ../foo`. If you choose to do this make sure to manually copy `py.typed` to the appropriate location, e.g. `venv/lib/python3.8/site-packages/foo`.

**Comment by tjsmart** [comment]
Thanks for the response and explanation! I'm still a bit surprised by the behavior but I'll read up on the documentation as to understand the choice.

Also, I can confirm that switching to `from ._bar import a as a` works! Wonder if there is a tool out there that can automate this for me 🤔 . Sounds like a fun little tool to write.

**Comment by erictraut** [comment]
I would recommend against automating this. It should be a deliberate decision by a library author to re-export a symbol from a module. By default, you probably want most imported symbols to remain private — an implementation detail of that module.

I agree that using a redundant form of an `import` looks weird. It was initially introduced by PEP 484 as a mechanism for [stub files](https://peps.python.org/pep-0484/#stub-files). The convention was adopted for "py.typed" libraries as well.

**Comment by tjsmart** [comment]
In the particular cases I am thinking of it was actually our intent for the symbol `a` to be public and visible to users of our package. It is our general strategy to have folders (e.g. `foo`) with private modules (e.g. `foo/_bar.py`) which we export specific symbols as public via the `__init__.py` file (e.g. `from ._bar import a`). So in our case the fact that a symbol being imported in the `__init__.py` file means we intend for it to be apart of our public interface.

With this strategy in mind, I think having a tool which ensures that symbols in our `__init__.py` which are intended to be public are considered public by tools like pylance would be nice to have.

But perhaps, we are creating our public interface in an unusual way, I'm always learning.
